keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,437 @@
+
+The algorithm to find the number of pods will look like this:
+
+```golang
+desiredReplicas = 0
+calculatedReplicas = AnyAlgorithmInHPAController(...)
+if calculatedReplicas > curReplicas:
+  constraint1 = CurReplicas * (1 + ScaleUpPercent/100)
+  constraint2 = CurReplicas + ScaleUpPods
+  scaleUpLimit = max(constraint1, constraint2)
+  desiredReplicas = min(scaleUpLimit, calculatedReplicas)
+else if calculatedReplicas < CurReplicas:
+  constraint1 = curReplicas * (1 - ScaleDownPercent/100)
+  constraint2 = CurReplicas - ScaleDownPods
+  scaleDownLimit = max(constraint1, constraint2)
+  desiredReplicas = max(scaleDownLimit, calculatedReplicas)
+```
+
+I.e., from the two provided constraints the larger one is used.
+
+If you don’t want to scale, you should set both parameters to zero for the appropriate direction (Up/Down).
+
+If only one parameter is given and the other is 0, then use the only defined constraint.
+
+The algorithm for `delay` will look like this:
+
+```golang
+for { // infinite cycle inside the HPA controller
+  desiredReplicas = calculateDesiredReplicas()
+  if desiredReplicas > curReplicas:
+    storeRecommendation(desiredReplicas, scaleUpRecommendations)
+    recommendations = getLastRecommendations(scaleUpRecommendations, ScaleUpDelaySeconds) // get recommendations for the last ScaleUpDelaySeconds
+    nextReplicas = min(recommendtaions)
+  if desiredReplicas < curReplicas:
+    storeRecommendation(desiredReplicas, scaleDownRecommendations)
+    recommendations = getLastRecommendations(scaleDownRecommendations, ScaleDownDelaySeconds) // get recommendations for the last ScaleDownDelaySeconds
+    nextReplicas = max(recommendtaions)
+  setReplicas(nextReplicas)
+  sleep(ControllerSleepTime)
+}
+```
+
+Effectively, this is a full copy of the current [Stabilization Window][] algorithm:
+
+- While scaling up, we should pick the safest (smallest) "desiredReplicas" number during last `delaySeconds`.
+- While scaling down, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+
+The “stabilization window" as a result becomes an alias for the `constraints.scaleDown.delaySeconds`.
+
+
+[Stabilization Window]: #stabilization-window
+
+#### Default Values
+
+For smooth transition it makes sense to set the following default values:
+
+- `constraints.scaleUp.delaySeconds = 0`, the delay is not used, instantly scale up
+- `constraints.scaleDown.delaySeconds = 300`, wait 5 min for the largest recommendation and then scale down to that value.
+- `constraints.scaleUp.rate.periodSeconds = 60`, one minute period for scaleUp
+- `constraints.scaleDown.rate.periodSeconds = 60`, one minute period for scaleDown
+- `constraints.scaleUp.rate.percent = 100`, allow to twice the number of pods
+- `constraints.scaleUp.rate.pods = 4`, i.e. allow adding up to 4 pods
+- `constraints.scaleDown.rate.percent = 100`, allow to remove all the pods
+- `constraints.scaleDown.rate.pods = math.MaxInt32`, allow to remove all the pods
```

Comment:
Default should be `nil`, not `math.MaxInt32`.

Reply:
Yep, good idea!
Fixed: https://github.com/kubernetes/enhancements/pull/883/commits/e47c961cbac46c4550830a7ea0f10a9c007641a9